### ISSUE TITLE
Memory leak in GraphConvTensorGraph

### ISSUE THREAD

**Issue by zealseeker** [deepchem/deepchem#1133]
Deepchem == 1.2.0
Python == 3.5.4

Repository to reproduce: https://github.com/zealseeker/deepchem-sandbox/tree/toxcast-loo

```python
for (train, test) in leave_one_out(len(dataset), valid_indice):
        start = time.clock()
        train_dataset = dataset.select(train, 'train_data')
        model = tf_graph_model(train_dataset, len(tasks), transformers)
        test_dataset = dataset.select(test,'test_data')
        y_preds.append(model.predict(test_dataset)[0][0])
        y_trues.append(list(next(test_dataset.itersamples()))[1])
```

I wanted to use "leave-one-out" to evaluate the performance of `GraphConvTensorGraph`. However, the memory used was continuing increasing so that I had to stop running.

In the first loop, the memory used was about 4GB. Then about 5GB, 6GB.. My computer has only 8GB while there were 1000+ models to build.

I tried to use `gc.collect`, only to find nothing changed.

**Comment by lilleswing** [comment]
So the way I am reading your code you are storing y_pred and y_true over and over in a loop
Toxcast has 8598 molecules over 618 tasks.  It is a categorical task so each task is one hot encoded to 2 integers.  An integer is 24 bytes in pure python.

```python
8598 * 618 * 24 * 2 * 2 / (1042* 1042) # ~487
# molecules * tasks * bytes per int * one_hot_encoding * y_pred, y_true / (bytes in megabyte)
```
So each loop you are loading ~ 500mb into RAM.
Flush y_pred and y_true to disk on each run.

**Comment by zealseeker** [comment]
Hi @lilleswing .  It is not caused by y_trues and y_preds I think. Each loop I just save the predictions. The final `y_preds` and `y_trues` should only occupy `1000 * 617 * 24 * 2` bytes, where 1000 is the number of compounds in "validation set".
Even if I tried removing the "y_true" and "y_pred" sentences, the issue still existed.

I'll try to use multiprocessing module to run each loop. (I still doubt that the model leads to memory leak)

**Comment by JanJorisDevogelaer** [comment]
Deepchem version: 2.3.0

I am experiencing a very similar issue, but in this case for the Keras version of GraphConvModel when performing a hyperparameter optimization with hyperopt. In a wrapper function, I initialize a the model, train for a certain amount of epochs and return several metrics to the optimizer. Subsequent runs of the optimization results in an accumulation of memory, which is interrupted with a ResourceExhaustedError after approximately 20-30 runs. 

I tried setting the variable in which the model is stored to None after fitting, but this does not seem have an effect. Any suggestions as to why the memory could keep on leaking?

Many thanks,
J

**Comment by rbharath** [comment]
@JanJorisDevogelaer Sorry for the late response, just seeing this comment.

@peastman Do you have any thoughts on what might be behind this memory leak?

**Comment by KacperKubara** [comment]
Hi,
I face exactly the same issue as @JanJorisDevogelaer. The GraphConv model leads to memory leakage while I use hyperopt or just a typical 10-fold CV (even if I explicitly delete the reference to the model with `del` keyword). After each run, RAM resources are depleted to the point when the program hangs.

It seems like the model is not explicitly deleted from the memory and all copy of the models just accumulate during the run.

**Comment by rbharath** [comment]
I'm taking a look at the `GraphConvModel` source code now. I don't see anything too exotic. There's just a keras model that's initialized. I think what's happening is that the underlying Keras model is sticking around in memory, depeleting the RAM of the memory.

I'm not sure there's an easy DeepChem workaround since this is a TensorFlow memory usage issue. The easiest fix might be to launch different models in different processes which can be killed outright afterwards. This is a little kludgy but should solve the issue I think

**Comment by KacperKubara** [comment]
Just a follow up. This seems to be related to this TF issue: https://github.com/tensorflow/tensorflow/issues/36465. They mention that it might be possible to clear the memory with `keras.backend.clear_session()` for TF 1.x

**Comment by rbharath** [comment]
Let me re-open this issue to track. I don't unfortunately have any leads on fixes here... When I get a chance I'll try training multiple graph convs on a single device and seeing if I can replicate this issue

**Comment by rbharath** [comment]
Closing this old issue since we don't see this on pytorch
